proposals/XXXX-spirv-resource-representation.md

```diff
@@ -0,0 +1,271 @@
+<!-- {% raw %} -->
+
+# HLSL resources in SPIR-V
+
+*   Proposal: [NNNN](NNNN-spirv-resource-representation.md)
+*   Author(s): [Steven Perron](https://github.com/s-perron)
+*   Status: **Design In Progress**
+
+*During the review process, add the following fields as needed:*
+
+*   PRs: [#114273](https://github.com/llvm/llvm-project/pull/114273),
+    [#111052](https://github.com/llvm/llvm-project/pull/111052),
+    [#111564](https://github.com/llvm/llvm-project/pull/111564),
+    [#115178](https://github.com/llvm/llvm-project/pull/115178)
+
+## Introduction
+
+There is a need to represent the HLSL resources in llvm-ir in a way that the
+SPIR-V backend is able to create the correct code. We have already done some
+implementation work for `Buffer` and `RWBuffer`. This was done as a
+proof-of-concept, and now we needed to determine how the other resource types
+will be represented.
+
+## Motivation
+
+The HLSL resources are fundamental to HLSL, and they are required in a Vulkan
+implementation.
+
+## Proposed solution
+
+We want to match the general solution proposed in
+[0006-resource-representations.md](0006-resource-representations.md). The
+`@llvm.spv.handle.fromBinding` intrinsic will be used to get a handle to the
+resource. It will return a target type to represent the handle. Then other
+intrinsics will be used to access the resource using the handle. Previous
+proposals left open what the target types should be for SPIR-V.
+
+The type for the handle will depend on the type of resource, and will be
+detailed in the following sections.
+
+The following sections will reference table 4 in the
+[shader resrouce interface](https://docs.vulkan.org/spec/latest/chapters/interfaces.html#interfaces-resources)
+for Vulkan.
+
+### SPIR-V target types
+
+The must be appropriate SPIR-V target types to represent the HLSL resources. We
+could try to represent the resources using the exact SPIR-V type that will be.
+The problem is that the HLSL resources does not map too closely with SPIR-V.
+
+Consider `StructuredBuffer`, `RWStructuredBuffer`,
+`RasterizerOrderedStructuredBuffer`, `AppendStructureBuffer`, and
+`ConsumeStructuredBuffer`. These resource types do not map directly to SPIR-V.
+They have multiple implicit features that need to map to different SPIR-V:
+
+1.  They all contains an array of memory that maps to a storage buffer.
+2.  Other than `StructuredBuffer`, they all contains a separate counter variable
+    that is its own storage buffer.
+3.  The references to `RasterizerOrderedStructuredBuffer` are contained in
+    implicit critical regions. In SPIR-V, explicit instructions are used to
+    start and stop the critical region.
+
+This makes it impossible to create a handle type that maps directly to a SPIR-V
+type. To handle this, we will create a target type `spv.VulkanBuffer`:
+
+```
+target("spv.VulkanBuffer", ElementType, StorageClass, IsWriteable, IsROV)
```

Comment:
```suggestion
target("spv.VulkanBuffer", ElementType, StorageClass, IsWriteable, IsROV [, CounterSet, CounterBinding])
```

Reply:
That's right (btw I'm assuming you did a type and meant `RWStructuredBuffer` for `c`)
But in that case, when you say `set & bindings can be provided by CounterSet/CounterBinding`, where would those be set?

Reply:
I missed removed those comments. I will do that.